subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -0,0 +1,2092 @@
+		atomic_clear_bit(ag->flags, BT_HFP_AG_CMEE_ENABLE);
+	} else {
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static int hfp_ag_update_indicator(struct bt_hfp_ag *ag, enum bt_hfp_ag_indicator index,
+				   uint8_t value, bt_hfp_ag_tx_cb_t cb, void *user_data)
+{
+	int error;
+	unsigned int key;
+	uint8_t old_value;
+
+	key = irq_lock();
+	old_value = ag->indicator_value[index];
+	if (value == old_value) {
+		LOG_ERR("Duplicate value setting, indicator %d, old %d -> new %d", index, old_value,
+			value);
+		irq_unlock(key);
+		return -EINVAL;
+	}
+
+	ag->indicator_value[index] = value;
+
+	irq_unlock(key);
+
+	LOG_DBG("indicator %d, old %d -> new %d", index, old_value, value);
+
+	error = hfp_ag_send_data(ag, cb, user_data, "\r\n+CIEV:%d,%d\r\n", (uint8_t)index + 1,
+				 value);
+	if (error) {
+		key = irq_lock();
+		ag->indicator_value[index] = old_value;
+		irq_unlock(key);
+	}
+
+	return error;
+}
+
+static void hfp_ag_close_sco(struct bt_hfp_ag *ag)
+{
+	LOG_DBG("");
+	if (NULL != ag->sco_chan.sco) {
+		bt_conn_disconnect(ag->sco_chan.sco, BT_HCI_ERR_LOCALHOST_TERM_CONN);
+		ag->sco_chan.sco = NULL;
+	}
+}
+
+static void bt_hfp_ag_reject_cb(struct bt_hfp_ag *ag, void *user_data)
+{
+	hfp_ag_close_sco(ag);
+	bt_hfp_ag_set_call_state(ag, BT_HFP_CALL_TERMINATE);
+
+	if (bt_ag && (bt_ag->reject)) {
+		bt_ag->reject(ag);
+	}
+}
+
+static void bt_hfp_ag_call_reject(struct bt_hfp_ag *ag, void *user_data)
+{
+	int error = hfp_ag_update_indicator(ag, BT_HFP_AG_CALL_SETUP_IND, BT_HFP_CALL_SETUP_NONE,
+					    bt_hfp_ag_reject_cb, user_data);
+	if (error != 0) {
+		LOG_ERR("Fail to send err :(%d)", error);
+	}
+}
+
+static void bt_hfp_ag_terminate_cb(struct bt_hfp_ag *ag, void *user_data)
+{
+	hfp_ag_close_sco(ag);
+	bt_hfp_ag_set_call_state(ag, BT_HFP_CALL_TERMINATE);
+
+	if (bt_ag && (bt_ag->terminate)) {
+		bt_ag->terminate(ag);
+	}
+}
+
+static void bt_hfp_ag_unit_call_terminate(struct bt_hfp_ag *ag, void *user_data)
+{
+	int error = hfp_ag_update_indicator(ag, BT_HFP_AG_CALL_IND, 0, bt_hfp_ag_terminate_cb,
+					    user_data);
+	if (error != 0) {
+		LOG_ERR("Fail to send err :(%d)", error);
+	}
+}
+
+static int bt_hfp_ag_chup_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err)
+{
+	int error;
+	bool ret;
+
+	*err = CME_ERROR_AG_FAILURE;
+
+	ret = is_char(buf, '\r');
+	if (ret != true) {
+		return -EINVAL;
+	}
+
+	if (ag->call_state == BT_HFP_CALL_ALERTING) {
+		if (!atomic_test_bit(ag->flags, BT_HFP_AG_INCOMING_CALL)) {
+			return -EINVAL;
+		}
+		error = hfp_ag_next_step(ag, bt_hfp_ag_call_reject, NULL);
+	} else if ((ag->call_state == BT_HFP_CALL_ACTIVE) || (ag->call_state == BT_HFP_CALL_HOLD)) {
+		error = hfp_ag_next_step(ag, bt_hfp_ag_unit_call_terminate, NULL);
+	} else {
+		return -EINVAL;
+	}
+	return error;
+}
+
+static uint8_t bt_hfp_get_call_state(struct bt_hfp_ag *ag)
+{
+	uint8_t status = 0;
+
+	switch (ag->call_state) {
+	case BT_HFP_CALL_TERMINATE:
+		break;
+	case BT_HFP_CALL_OUTGOING:
+		status = 2;
```

Comment:
Wouldn't it make sense to have some defines for these status values, rather than just having hardcoded values here?

Reply:
updated